### PR TITLE
[coordinator] Split NewConnectionFn into NewClientFn

### PR DESCRIPTION
**Pull request by Antanukas**
**What this PR does / why we need it**:

Refactors code to reduce probability of introducing channel leak in the future.

**Special notes for your reviewer**:

Refactoring code related to opening a TChannel and constructing rpc client. Both session and connection_pool will reuse `establishConnection`.

1. Adds `NewClientFn`, `HealthCheckNewConnFn`, `HealthCheckFn` to be able to have a finer grained configuration of connection pool.
2. `NewConnectionFn` is now responsible to create TChannel only. Client creation moved to  `NewClientFn`.
3. `establishNewConnection` private func introduced to cover full lifecycle of creating new connection and ensuring channel is closed in case of failure.
4. New connection errors are now logged with `Warn` instead of debug. It should not be spammy since pool tries to reconnect everything every 4 seconds.

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/dbnode/client/connection_pool.go [hunk 1]
```diff
@@ -182,30 +182,25 @@ func (p *connPool) connectEvery(interval time.Duration, stutter time.Duration) {
...
 				}
 
 				p.maybeEmitHealthStatus(healthStatusOK)
 				p.Lock()
 				if p.status == statusOpen {
-					p.pool = append(p.pool, conn{channel, client})
+					p.pool = append(p.pool, conn{ch, client})
 					p.poolLen = int64(len(p.pool))
+				} else {
+					ch.Close()
```
Comment (Antanukas) [inline src/dbnode/client/connection_pool.go:203]: This is likely a corner case since pool is being closed on shutdown only. Just trying to be a good citizen and cleanup. Also did not figure an easy way to test this.

File: src/dbnode/client/options.go [hunk 2]
```diff
@@ -327,26 +328,6 @@ func NewOptionsForAsyncClusters(opts Options, topoInits []topology.Initializer,
 	return result
 }
 
-func defaultNewConnectionFn(
```
Comment (Antanukas) [inline src/dbnode/client/options.go:330]: moved to connection_pool.go and split into 2 methods: NewConnectionFn and NewClientFn

File: src/dbnode/client/session.go [hunk 3]
```diff
@@ -777,18 +775,12 @@ func (s *session) DedicatedConnection(
...
-		newConnFn := s.opts.NewConnectionFn()
-		channel, client, err = newConnFn(channelName, host.Address(), s.opts)
+		channel, client, _, err = establishNewConnection(host.Address(), opts.BootstrappedNodesOnly, s.opts)
 		if err != nil {
 			multiErr = multiErr.Add(err)
 			return
 		}
 
-		if err := s.healthCheckNewConnFn(client, s.opts, opts.BootstrappedNodesOnly); err != nil {
-			multiErr = multiErr.Add(err)
```
Comment (Antanukas) [inline src/dbnode/client/session.go:788]: here we were missing a channel close statement. Now this is fixed by reusing `establishNewConnection`

**Comment by linasm** [comment]
Could we just add the missing `close` instead of a whole rewrite of critical piece of code? Or at least do it in 2 separate PRs?

**Comment by Antanukas** [comment]
If I split to separate PRs it would require double work - I will need to implement additional test suite in `session` and in `connection_pool.go`. By doing this refactoring I moved testing to `establishConnection` and removed responsibility of closing a channel in session.go

File: src/dbnode/client/connection_pool.go [hunk 4]
```diff
@@ -182,34 +182,27 @@ func (p *connPool) connectEvery(interval time.Duration, stutter time.Duration) {
 		var wg sync.WaitGroup
 		for i := 0; i < target-poolLen; i++ {
 			wg.Add(1)
-			newConnFn := p.opts.NewConnectionFn()
 			go func() {
 				defer wg.Done()
 
-				// Create connection
-				channel, client, err := newConnFn(channelName, address, p.opts)
```
Comment (ryanhall07) [inline src/dbnode/client/connection_pool.go:190]: how was the channel being leaked previously?

Comment (ryanhall07) [inline src/dbnode/client/connection_pool.go:190]: i don't follow what the original leak was

**Comment by Antanukas** [comment]
I did split this PR into smaller one with channels closed https://github.com/m3db/m3/pull/4026

File: src/dbnode/client/connection_pool.go (hunk 4 above)
Comment (Antanukas) [reply src/dbnode/client/connection_pool.go:190]: Here there was no leak. I have extracted code into establishConnection and reused it for dedicated session that had a leak before I fixed it here https://github.com/m3db/m3/pull/4026
